File: randomProblems/piday.cpp
#include <iostream>

using namespace std;

int possibleAfter(int k, int n, int currentSum, int currentLength, int lastElement) {
	if (currentLength == k - 1) {
		if (currentSum + lastElement <= n)
			return 1;
		else
			return 0;
	}

	int max_possible = (n - currentSum) / (k - currentLength);

	int min_possible;
	if (currentLength == 0)
		min_possible = 1;
	else {
		if (max_possible < lastElement)
			return 0;

		if (currentLength == k - 1)
			min_possible = n - currentSum;
		else
			min_possible = lastElement;
	}

	int possibilities = 0;
	for (int i = min_possible; i <= max_possible; i++)
		possibilities += possibleAfter(k, n, currentSum + i, currentLength + 1, i);

	return possibilities;
}

int main() {
	int n;
	int k;
	cin >> n;
	cin >> k;

	if (n == k || k == 1)
		cout << 1 << endl;
	else
		cout << possibleAfter(k, n, 0, 0, 0) << endl;;

	return 0;
}

File: randomProblems/piday.py
import sys
n = int(sys.stdin.readline())  # pies
k = int(sys.stdin.readline())  # people

cache = {}


def possibleAfter(currentSum, currentLength, lastElement):
	if (currentSum, currentLength, lastElement) in cache:
		return cache[(currentSum, currentLength, lastElement)]

	if currentLength == k:
		if currentSum < n:
			return 0
		else:
			return 1 

	stepsLeft = k - currentLength
	piesLeft = n - currentSum

	max_possible = int(piesLeft / stepsLeft)  # problem here probably

	if currentLength == 0:
		min_possible = 1
	else:
		if max_possible < lastElement:
			return 0

		if currentLength == k - 1:
			min_possible = piesLeft
		else:
			min_possible = lastElement

	# print(min_possible, max_possible)

	possibilities = 0
	for i in range(min_possible, max_possible + 1):
		possibilities += possibleAfter(currentSum + i, currentLength + 1, i)

	cache[(currentSum, currentLength, lastElement)] = possibilities
	return possibilities


print(possibleAfter(0, 0, 0))

